alica_engine/src/engine/BasicBehaviour.cpp

```diff
@@ -113,30 +100,36 @@ void BasicBehaviour::doTrigger()
         return;
     }
     _triggeredJobRunning = true;
-    _engine->editScheduler().schedule(std::bind(&BasicBehaviour::runJob, this, nullptr));
+    _engine->editScheduler().schedule(std::bind(&BasicBehaviour::runJob, this));
 }
 
-void BasicBehaviour::doTerminate()
+void BasicBehaviour::terminateJob()
 {
-
     if (_activeRunJobId != -1) {
         _engine->editScheduler().cancelJob(_activeRunJobId);
         _activeRunJobId = -1;
     }
     // Just to be double safe in terms of the correct behaviour of isSuccess() & isFailure() ensure result is reset before incrementing _execState
     _behResult.store(BehResult::UNKNOWN);
-    setTerminatedState();
-
-    // Intentionally call onTermination() at the end. This prevents setting success/failure from this method
-    try {
-        traceTermination();
-        onTermination();
-    } catch (const std::exception& e) {
-        ALICA_ERROR_MSG("[BasicBehaviour] Exception in Behaviour-TERMINATE of: " << getName() << std::endl << e.what());
+    doTerminate();
+}
+
+void BasicBehaviour::onInit_()
```

Comment:
Or maybe just reverse the "dependency" of the functions

have a `doInitGeneric` that contains the shared logic, then a virtual `doInit` that calls the generic function, then does a little bit of behavior/plan specific logic.  That would be much better imo than this, where we have a generic function that calls an overridden specific function.  Either way I think we could get rid of these three functions.

Reply:
I couldn't find another way to do this because of the termination code in doTerminate() which is hard to separate cleanly. Anyway the reasoning behind the current implementation is: the BasicBehaviour/BasicPlan objects implement & manipulate (schedule/cancel) init, run & terminate jobs which use the Runnable object's init, run & termination (doInit, doRun & doTerminate) logic but override what happens once the init, run, terminate is actually done i.e. onInit_, onRun_ & onTerminate_

Reply:
is it okay if I get rid of onInit_ & onRun_ but retain onTerminate_?

Reply:
Maybe we can at least separate the scheduling functions from the JIT functions.  `start` could be renamed `schedule`

scheduleSomething() is called when scheduling and `doSomething` is what is called when actually doing it.  Then we think of a way to make it more clear from the naming which function calls which or which function is the generic one

Reply:
its already that way except for the start/stop functions. Please see above comment for an explanation